### DISCUSSION TITLE
PIP-25: Dynamic fees structure

### DISCUSSION THREAD

**Discussion by ja7ad** [pactus-project/PIPs#117] [PIPs]
PIP-25: [Dynamic fees structure](https://pips.pactus.org/PIPs/pip-25)

**Comment by b00f** [comment]
Please explain how to calculate:
"Current Transaction Volume" and "Average Transaction Volume"

The current implementation we have is based on the number of unconfirmed transactions inside the  pool. Each node can decide how to set the base fee and the size for the pool.

**Reply by ja7ad** [reply]
The "Current Transaction Volume" is a measure of the number of transactions waiting to be included in the blockchain at any given moment. Since your implementation is based on the number of unconfirmed transactions inside the pool, you can calculate it as follows:

1. **Monitor the Mempool**: Each node monitors the number of transactions in its transaction pool (mempool).
2. **Count Unconfirmed Transactions**: Sum the total number of unconfirmed transactions in the mempool.

**Calculation Steps**:
- **Step 1**: Access the mempool data.
- **Step 2**: Count the number of transactions currently in the mempool.

**Formula**:

```math
\[ \text{Current Transaction Volume} = \text{Number of unconfirmed transactions in the mempool} \]
```

The "Average Transaction Volume" represents the typical number of transactions waiting to be included in a block over a certain period or number of blocks. This provides a baseline to determine what is considered "normal" activity on the network.

1. **Define a Window**: Choose a time window or a specific number of blocks to calculate the average. For example, you could use the last 100 blocks.
2. **Collect Historical Data**: For each block in the chosen window, record the number of transactions that were waiting to be included.
3. **Calculate the Average**: Compute the average number of transactions over the selected blocks.

**Calculation Steps**:
- **Step 1**: Define the observation window (e.g., last 100 blocks).
- **Step 2**: For each block in the window, record the number of transactions that were included from the mempool.
- **Step 3**: Calculate the average of these recorded numbers.

**Formula**:

```math
\[ \text{Average Transaction Volume} = \frac{\sum_{i=1}^{N} \text{Transaction Volume in Block } i}{N} \]
```

where $\( N \)$ is the number of blocks in the observation window.

**Example**

Assume you want to use the last 100 blocks to calculate the average transaction volume.

**Pseudocode Example**

```python
# Parameters
OBSERVATION_WINDOW = 100  # Number of blocks to consider for the average

def get_current_transaction_volume(mempool):
    # Count the number of unconfirmed transactions in the mempool
    return len(mempool)

def get_average_transaction_volume(blockchain, current_block_height, observation_window=OBSERVATION_WINDOW):
    # Collect transaction volumes from the last observation_window blocks
    total_transactions = 0
    for i in range(current_block_height - observation_window, current_block_height):
        block = blockchain.get_block_by_height(i)
        total_transactions += len(block.transactions)
    
    # Calculate the average
    average_transaction_volume = total_transactions / observation_window
    return average_transaction_volume

# Example usage
current_mempool = get_mempool()  # Function to get the current mempool
blockchain = get_blockchain()    # Function to access the blockchain data
current_block_height = blockchain.get_current_height()

current_transaction_volume = get_current_transaction_volume(current_mempool)
average_transaction_volume = get_average_transaction_volume(blockchain, current_block_height)

print("Current Transaction Volume:", current_transaction_volume)
print("Average Transaction Volume:", average_transaction_volume)
```

**Dynamic Fee Adjustment Using These Volumes**

Once you have the current and average transaction volumes, you can adjust the base fee dynamically:

```python
BASE_FEE = 0.01  # Example base fee in tokens
ALPHA = 0.05     # Sensitivity coefficient

def calculate_adjusted_fee(base_fee, alpha, current_volume, average_volume):
    adjustment_factor = 1 + alpha * (current_volume / average_volume)
    adjusted_fee = base_fee * adjustment_factor
    return adjusted_fee

adjusted_fee = calculate_adjusted_fee(BASE_FEE, ALPHA, current_transaction_volume, average_transaction_volume)
print("Adjusted Fee:", adjusted_fee)
```
